wagtail/wagtailcore/models.py

```diff
@@ -1377,11 +1378,64 @@ def serve_password_required_response(self, request, form, action_url):
         context['action_url'] = action_url
         return TemplateResponse(request, self.password_required_template, context)
 
+    def is_permitted_root(self, request, choosable=False):
+        """
+        Returns True if this page is the root of the current user's permitted section of the page tree.
+        """
+        # A superuser's permitted root is the Root page.
+        if request.user.is_superuser:
+            return self.is_root()
+
+        return self.path == get_closest_common_ancestor_path(request, choosable)
+
+    def get_permitted_ancestors(self, request, choosable=False):
+        """
+        Returns the permitted ancestors for a user on a given request.
+        """
+        # Superusers see all the ancestors.
+        if request.user.is_superuser:
+            return self.get_ancestors()
+
+        # Copied this algorithm from TreeQuerySet.ancestor_of_q, but tweaked it to remove this page's path.
+        ancestor_paths = [
+            self.path[0:pos]
+            for pos in range(0, len(self.path) + 1 - self.steplen, self.steplen)[1:]
+        ]
+
+        cca_path = get_closest_common_ancestor_path(request, choosable)
+        if cca_path:
+            # Return only pages with paths that are at least as long as the Closest Common Ancestor's path.
+            explorable_ancestor_paths = [path for path in ancestor_paths if len(path) >= len(cca_path)]
+            return Page.objects.filter(path__in=explorable_ancestor_paths)
+        else:
+            # If the user has no permitted paths, they have no explorable ancestors.
+            return Page.objects.none()
+
     class Meta:
         verbose_name = _('page')
         verbose_name_plural = _('pages')
 
 
+def get_closest_common_ancestor_path(request, choosable=False):
+    """
+    Global method that returns the Closest Common Ancestor for the specified User, or None if the User has no
+    explorable Page paths. Pass in choosable=True to get the CCA path of the User's choosable pages, rather
+    then their explorable pages.
+    """
```

Comment:
As of #3072 this functionality now exists as part of PageQuerySet <https://github.com/torchbox/wagtail/blob/cc299d8873f1dc02e99d4fffee5b670d4c00fd15/wagtail/wagtailcore/query.py#L235>, so I'll see if we can re-use that rather than repeating the logic.